### PR TITLE
net/mlx5: separate mlx5_rx_burst into a regular vs optimized no_sges versions (ETH and IPoIB)

### PR DESCRIPTION
**Pull request by tomerfiliba**
Backport mlx5_rx_replenish_bulk_mbuf from DPDK 17.11

### CODE REVIEW

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 1]
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
 	return ol_flags;
 }
 
+#define MLX5_VPMD_RXQ_RPLNSH_THRESH  (64U)
+
+/**
+ * Replenish buffers for RX in bulk.
+ * Backported from DPDK 17.11
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: no need for the backport comment

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 2]
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
...
+ */
+static inline void
+mlx5_rx_replenish_bulk_mbuf(struct rxq *rxq, signed int rel_data_offset) {
+	const uint16_t q_n = 1 << rxq->elts_n;
+	const uint16_t q_mask = q_n - 1;
+	uint16_t elts_idx = rxq->rq_ci & q_mask;
+	unsigned int i;
+
+	assert(rxq->refill_count >= MLX5_VPMD_RXQ_RPLNSH_THRESH);
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c:2264]: redundent line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 3]
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
...
+	while (rxq->refill_count > 0) {
+		struct rte_mbuf **elts = &(*rxq->elts)[elts_idx];
+		volatile struct mlx5_wqe_data_seg *wq = &(*rxq->wqes)[elts_idx];
+		/* Not to cross queue end. */
+		uint16_t batch_size = RTE_MIN(rxq->refill_count, q_n - elts_idx);
+
+		if (rte_mempool_get_bulk(rxq->mp, (void *) elts, batch_size) < 0) {
+			//RTE_LOG(ERR, PMD, "PMD: RX ring couldn't be replenished, missing %d pkts (batch_size=%d)\n",
+			//        rxq->refill_count, batch_size);
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: remove the commented lines

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 4]
```diff
@@ -2242,6 +2243,51 @@ rxq_cq_to_ol_flags(struct rxq *rxq, volatile struct mlx5_cqe *cqe)
...
+#define MLX5_VPMD_RXQ_RPLNSH_THRESH  (64U)
+
+/**
+ * Replenish buffers for RX in bulk.
+ * Backported from DPDK 17.11
+ *
+ * @param rxq
+ *   Pointer to RX queue structure.
+ * @param rel_data_offset
+ *   Should be 0 for ETH, GRH_HDR_LEN for IPoIB
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: need to explain what this value means, and not with what values it should be set.

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 5]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+	unsigned int i = 0;
+	unsigned int rq_ci = (rxq->rq_ci) + rxq->refill_count;
+	unsigned int first_idx = rq_ci & wqe_cnt;
+	const unsigned char port = rxq->port_id;
+	int len = 0; /* keep its value across iterations. */
+
+	rte_prefetch0(cqe);
+	rte_prefetch0(cqe+1);
+	rte_prefetch0(cqe+2);
+	// prefetch the next cache line (no point in prefetching the current)
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: comment style is wrong.

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 6]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+	while (pkts_n) {
+		unsigned int idx = rq_ci & wqe_cnt;
+		struct rte_mbuf *rep = (*rxq->elts)[idx];
+		unsigned int next_idx = (idx + 1) & wqe_cnt;
+		uint32_t rss_hash_res = 0;
+		cqe = &(*rxq->cqes)[rxq->cq_ci & cqe_cnt];
+
+		if ((idx & 4) == 0) {
+			struct rte_mbuf **next_mbufs_cache_line = &(*rxq->elts)[(idx + (64/sizeof(struct rte_mbuf *))) & wqe_cnt];
+			// we start a new cache line, prefetch the next one
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: comment style is wrong. there are more cases below

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 7]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+			rte_prefetch0(next_mbufs_cache_line);
+		}
+		rte_prefetch0(cqe+3); // we're allowed to prefetch invalid addresses
+		rte_prefetch0((*rxq->elts)[next_idx]); // prefetch the next mbuf header
+		seg = rep;
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 8]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+		if (!len) {
+			break;
+		}
+		if (unlikely(len == -1)) {
+			RTE_LOG(ERR, PMD, "PMD: len is -1\n");
+			/* RX error, packet is likely too large. */
+			++rxq->stats.idropped;
+			goto skip;
+		}
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c:2511]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 9]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+		struct rte_mbuf *pkt = seg;
+		assert(len >= (rxq->crc_present << 2));
+		/* Update packet information. */
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 10]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+		SET_DATA_OFF(pkt, RTE_PKTMBUF_HEADROOM);
+		NB_SEGS(pkt) = 1;
+		PORT(pkt) = port;
+		NEXT(pkt) = NULL;
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 11]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+		if (rxq->vlan_strip &&
+			(cqe->hdr_type_etc &
+			 htons(MLX5_CQE_VLAN_STRIPPED))) {
+			pkt->ol_flags |= PKT_RX_VLAN_PKT |
+				PKT_RX_VLAN_STRIPPED;
+			pkt->vlan_tci = ntohs(cqe->vlan_info);
+		}
+		if (rxq->crc_present)
+			len -= ETHER_CRC_LEN;
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c:2552]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 12]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+		// sges_n == 0 so these two are equal
+		PKT_LEN(pkt) = len;
+		DATA_LEN(pkt) = len;
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 13]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+		*(pkts++) = pkt;
+		--pkts_n;
+		++i;
+skip:
+		/* Align consumer index to the next stride. */
+		++rq_ci;
+	}
+	if (unlikely((i == 0) && (rq_ci == rxq->rq_ci)))
+		return 0;
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: redundant blank line

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 14]
```diff
@@ -2402,6 +2448,145 @@ mlx5_rx_burst_eth(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+	rxq->refill_count += i;
+	if (rxq->refill_count >= MLX5_VPMD_RXQ_RPLNSH_THRESH) {
+        // RTE_LOG(ERR, PMD, "PMD: will refill %d pkts starting at %d. first_idx=%d i=%d rq_ci=%d wqe_cnt=%d\n",
+        //         rxq->refill_count, rxq->rq_ci, first_idx, i, rq_ci, wqe_cnt);
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: remove commented lines

File: drivers/net/mlx5/mlx5_rxtx.c [hunk 15]
```diff
@@ -2538,6 +2723,117 @@ mlx5_rx_burst_ipoib(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n)
...
+ * @param pkts_n
+ *   Maximum number of packets in array.
+ *
+ * @return
+ *   Number of packets successfully received (<= pkts_n).
+ */
+uint16_t
+mlx5_rx_burst_ipoib_no_sges(void *dpdk_rxq, struct rte_mbuf **pkts, uint16_t pkts_n) {
+	assert(rxq->sges_n == 0);
+
```
Comment (shahafsh) [inline drivers/net/mlx5/mlx5_rxtx.c]: same comment as in eth burst function.

redundant extra lines.
wrong comment style.

**Review by shahafsh — COMMENTED** [review]
Hi Tomer,

i added few comment regarding to coding style.
even though this is not upstream code i would like to keep on high standard, for us to be able to maintain it easily.

my big question is -
i understand the big improvement by replenish set of buffers in bulk.

however how much gain you see when not using sge in compare to the version which does?

if it is not large, i suggest to modify the current eth and ipoib burst to replenish buffers in bulk and not introduce new set of data path functions.

**Comment by shahafsh** [comment]
@tomerfiliba 

please see some comments and question

**Comment by tomerfiliba** [comment]
@shahafsh, the replenish logic only works if `sges_n == 0`. this is also the case in the vectorized version we took it from

File: drivers/net/mlx5/mlx5_rxtx.c (hunk 2 above)
Comment (tomerfiliba) [reply drivers/net/mlx5/mlx5_rxtx.c:2264]: the assert? we took it from http://dpdk.org/browse/dpdk/tree/drivers/net/mlx5/mlx5_rxtx_vec.h?h=v17.11#n109

**Comment by tomerfiliba** [comment]
@shahafsh i removed most of the blank lines and fixed the comments, however, some of the blank lines are there for readability (marking the boundaries of code blocks)

**Comment by shahafsh** [comment]
@tomerfiliba 

squashed the CR fix and push to master branch with some commit log rewroding.

thanks.

**Pull request closed without merging**